### ISSUE TITLE
No indication a call to a AWS lambda upstream is being throttled by AWS.

### ISSUE THREAD

**Issue by chris530** [kgateway-dev/kgateway#10192]
# Context

User's Lambda's can be throttled, which results in the Lambda returning a 429 response code (i.e. an actual 429 - Too Many Requests, not a 429 wrapped in a 200). When this happens, and you have `unwrapAsApiGateway` enabled, Gloo sends a 200 back to the client with an empty response, and a 200 is logged in the Gloo Prometheus metrics, providing no indication that the lambda request was throttled.

See the following comment for details, and an explanation of the AWS APIGateway semantics in this case: https://github.com/k8sgateway/k8sgateway/issues/10192#issuecomment-2470437732

After presenting the user with our findings, we've come to the conclusion that we should have the same semantics as AWS APIGateway in this case (i.e. return a 500), but the user would like us to set additional response flags that show the original error from the upstream lambda. See this comment: https://github.com/k8sgateway/k8sgateway/issues/10192#issuecomment-2567674273

### Gloo Edge Product
Enterprise

### Gloo Edge Version
v1.17.3

### Is your feature request related to a problem? Please describe.
I can create a upstream / vs to an AWS lambda function, by also enabling unwrapAsApiGateway  I see the desired code in the response.    My issue is when AWS is throttling the function,   the response code always comes back as 200.

## ```
apiVersion: gateway.solo.io/v1
kind: VirtualService
metadata:
  name: nginx
  namespace: gloo-system
spec:
  virtualHost:
    domains:
    - 'nginx.web'
    routes:
    - matchers:
      - exact: /
      routeAction:
        single:
          destinationSpec:
            aws:
              logicalName: christest
              unwrapAsApiGateway: true
          upstream:
            name: aws-upstream

##             namespace: gloo-system
apiVersion: gloo.solo.io/v1
kind: Upstream
metadata:
  name: aws-upstream
  namespace: gloo-system
spec:
  aws:
    region: us-east-2
    secretRef:
      name: aws-creds
      namespace: gloo-system

```

### Describe the solution you'd like

The response is logged to the prometheus metric  envoy_cluster_external_upstream_rq which is then reflected to the grafana dashboard.  

Instead of a 200,   if the lambda function is being throttled,  I believe a 429 should be returned instead,  which can be  then scraped by prometheus. 

### Describe alternatives you've considered

Enabling the envoy AWS debug logs,   I have looked for any indication that the throttled is happening.  I can't find  any evidence of this from envoy. 

The AWS UI does display the function is being throttled  and if I execute the function via the aws cli,  a exception is thrown as expected  "TooManyRequestsException" . 

### Additional Context

Envoy in debug mode.   Not throttled.

Lambda 

```

import json

def lambda_handler(event, context):
    # Create a response with a 410 status code
    response = {
        "statusCode": 410,
        "body": json.dumps({
            "message": "Too Many Requests Chris"
        }),
        "headers": {
            "Content-Type": "application/json"
        }
    }

```
return response
```

```

```

gateway-proxy-5b864f8f88-9jrfw gateway-proxy ':status', '410'
gateway-proxy-5b864f8f88-9jrfw gateway-proxy ':status', '410'
gateway-proxy-5b864f8f88-9jrfw gateway-proxy ':status', '410'
gateway-proxy-5b864f8f88-9jrfw gateway-proxy ':status', '410'
gateway-proxy-5b864f8f88-9jrfw gateway-proxy ':status', '410'
gateway-proxy-5b864f8f88-9jrfw gateway-proxy ':status', '410'
gateway-proxy-5b864f8f88-9jrfw gateway-proxy ':status', '410'

```
Envoy in debug mode.  throttled.

```

gateway-proxy-5b864f8f88-9jrfw gateway-proxy ':status', '200'
gateway-proxy-5b864f8f88-9jrfw gateway-proxy ':status', '200'
gateway-proxy-5b864f8f88-9jrfw gateway-proxy ':status', '200'
gateway-proxy-5b864f8f88-9jrfw gateway-proxy ':status', '200'
gateway-proxy-5b864f8f88-9jrfw gateway-proxy ':status', '200'
gateway-proxy-5b864f8f88-9jrfw gateway-proxy ':status', '200'
gateway-proxy-5b864f8f88-9jrfw gateway-proxy ':status', '200'

```

I did find this in the envoy docs.

_The status_code field is an integer used as the HTTP response code. If this key is missing, Envoy returns a 200 OK._

Which may be why it's receiving the 200.



┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1206681874205238/1208510335689301) by [Unito](https://www.unito.io)

**Comment by nfuden** [comment]
Weve made some pretty big changes recently that may let us do this more cleanly. 

From a technical space we say that unwrapasapigateway will neatly unwrap and treat the request as if it was never in the api gateway format. Part of the apigateway format is that if you managed to reach aws then you will always [get a 200](https://docs.aws.amazon.com/apigateway/latest/developerguide/handle-errors-in-lambda-integration.html) and then have the inner piece with whatever internal error wrapped. 


In other words what we need to do is move unwrapas functionality from its current state in the filter chain into the new upstream filters location.

**Comment by DuncanDoyle** [comment]
@nfuden So that would basically mean that, because we unwrap earlier in the chain, the response from the upstream is the response that was wrapped in the AWS Lambda Response Event format (https://docs.aws.amazon.com/lambda/latest/dg/services-apigateway.html#apigateway-types-transforms), correct?

If we change that behaviour, is there anything that could semantically break for existing users of "unwrapAsApiGateway"?

**Comment by nfuden** [comment]
The impact here is if there are any filters after aws transformer and before the hcm that materially change something. This would most likely be seen in another instance of a transformation filter.

And yes basically instead of unwrapping later in the chain we are doing it up front so that all the stats and  things in the hcm treat the request as its final form rather than the pre transform variant.

**Comment by BeauSelisker28** [comment]
@DuncanDoyle --- @chris530 or @kevin-shelaga will aim to weigh in regarding your questions. We may need to circle up with Carfax for a response.

**Comment by nfuden** [comment]
Nice investigation @DuncanDoyle ! I see I was wrong about the 429 format from aws thanks for digging in.

I would argue that this feature is meant to remove as much of the `oddness` of apigateway. Given this I think we want to augment lambda unwrap to inspect the response reasons from x-amzn-errortype and decide on appropriate response codes to surface to the caller.

**Comment by BeauSelisker28** [comment]
@nfuden Looks like we seem to have made some progress on determining our path forward. What's your estimation on the LOE here and a potential timeline on when we might be able to bring this forward? Carfax is keen to get some sense of timing. Thanks!

**Comment by chris530** [comment]
My two cents,  if the lambda / apigw  returns a 429 ,  that should be the response code sent back to the client.

**Comment by DuncanDoyle** [comment]
But the point is that apigw DOES NOT send a 429 back to the client if the Lambda is throttled. The Lambda sends a 429 (so not a 429 wrapped in a 200, but an actual HTTP 429) back to the Gateway, and AWS API Gateway sends a 500 back to the client (as explained in my comment [here](https://github.com/k8sgateway/k8sgateway/issues/10192#issuecomment-2470437732)).

**Comment by DuncanDoyle** [comment]
After a call with the customer, we've come to the agreement that Gloo Gateway's Lambda `unwrapAsApiGateway` should behave the same as AWS API Gateway when dealing with 429 responses coming from upstream Lambda's, i.e. send a 500 to the client.

However, we should add a response flag to our response (e.g. response header) that contains the original response code from the upstream Lambda, in this case a 429.

**Comment by nfuden** [comment]
Copied to https://github.com/solo-io/solo-projects/issues/7827 for Solo customer uses.
There is an open question as to whether kgateway will want to use the envoy-gloo aws lambda or try to standardize on upstream envoy lambda functionality along with custom transformers reimplemented in rustformations (tbd for envoy-gloo parity)

**Comment by ashishb-solo** [comment]
I've posted a fix for this issue on https://github.com/solo-io/envoy-gloo/pull/407

However, I would like to be absolutely sure that I'm doing the correct thing. So to recap this thread, my understanding is as follows:

IF
  aws returns a status code 429 (in the header, not the json_body.status_code field)
AND
  unwrapAsApiGateway == true
THEN
  end users should receive a response code of 500
AND
  we should set a header (i chose to call it `x-envoy-lambda-statuscode`) to 429

Additionally, we ignore the body of the response from AWS and do not set any of the headers that we typically set (statuscode, headres, multiValueHeaders, etc.). Also, note that this header is only set when AWS returns a 429, not for any other responses.

In other words, with the fix applied, this is how curl requests look in my setup:

```
$ curl -sv localhost:8080 -d '{}' -X POST
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> POST / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.88.1
> Accept: */*
> Content-Length: 2
> Content-Type: application/x-www-form-urlencoded
> 
< HTTP/1.1 500 Internal Server Error
< x-envoy-lambda-statuscode: 429
< content-length: 2
< date: Tue, 11 Mar 2025 00:27:56 GMT
< server: envoy
< 
* Connection #0 to host localhost left intact
{}
```

@DuncanDoyle I would appreciate if you could make sure that I'm understanding everything correctly and that the fix works as intended. Thanks!

**Comment by ashishb-solo** [comment]
https://github.com/solo-io/gloo/pull/10701 merged. At this point, I think it's safe to close this.
